src/measure_theory/borel_hierarchy.lean

```diff
@@ -0,0 +1,529 @@
+    rw mem_range,
+    have typf : ∀ n, ↑(f n) ∈ ⋃ (j < order.succ (sup o)), pi0 s j,
+    { intro n, apply mem_Union.2,
+      specialize ho n,
+      use o n,
+      exact mem_Union.2 ⟨lt_of_le_of_lt (le_sup o n) (order.lt_succ (sup o)), ho.2⟩ },
+    use λ n, (⟨f n, typf n⟩ : ⋃ (j < order.succ (sup o)), pi0 s j),
+    tauto },
+  { simp only [Union_subset_iff],
+    exact λ _ hi, sigma0_subset_sigma0 s _ _ (le_of_lt hi) }
+end
+
+theorem compl_mem_gen_measurable (t : set α) (ht : t ∈ gen_measurable s) :
+  tᶜ ∈ gen_measurable s :=
+begin
+  rw gen_measurable_eq_Union_sigma0 at ht,
+  simp only [mem_Union,exists_prop] at ht,
+  cases ht with o ho,
+  rcases classical.em (o=0) with rfl | onon,
+  { finish },
+  calc
+  tᶜ  ∈ pi0 s o          : by { rw pi0_eq_compl_sigma0,
+    simp only [mem_image,compl_inj_iff,exists_eq_right], exacts [ho.2,onon] }
+  ... ⊆ gen_measurable s : pi0_subset_sigma0 s o ω₁ ho.1,
+end
+
+theorem Union_mem_gen_measurable {f : ℕ → set α} (hf : ∀ n, f n ∈ gen_measurable s) :
+  (⋃ n, f n) ∈ gen_measurable s:=
+by { unfold gen_measurable at *, exact Union_of_mem_sigma0 s ω₁ hf }
+
+open measurable_space
+
+lemma generate_measurable_of_mem_sigma0 (t) (ht : t ∈ sigma0 s i) :
+  generate_measurable s t :=
+begin
+  induction i using ordinal.induction with i IH generalizing t,
+  rw [sigma0_eq_Union_pi0,mem_range] at ht,
+  rcases ht with ⟨f,hf⟩,
+  have typf : ∀ n : ℕ, generate_measurable s (f n),
+  { intro n,
+    have fn_in : (f n).val ∈ ⋃ (j : ordinal) (hij : j < i), pi0 s j := (f n).property,
+    simp only [subtype.val_eq_coe,mem_Union,exists_prop] at fn_in,
+    rcases fn_in with ⟨o,⟨o_lt_i,fn_in⟩⟩,
+    -- Case `(f n).val ∈ pi0 s 0`.
+    rcases classical.em (o=0) with rfl | honz,
+    { rw pi0_zero at fn_in,
+      rcases fn_in with  fn_in | fn_emp | fn_in,
+      { exact generate_measurable.basic _ fn_in },
+      { rw fn_emp, exact generate_measurable.empty },
+      { rw mem_singleton_iff at fn_in, rw [fn_in,←compl_empty],
+        exact generate_measurable.compl _ generate_measurable.empty } },
+    -- Case `(f n).val ∈ pi0 s o` with `o ≠ 0`.
+    simp only at IH,
+    rw pi0_eq_compl_sigma0 s o honz at fn_in,
+    rw ← compl_compl ↑(f n),
+    apply generate_measurable.compl,
+    cases fn_in with x hx,
+    rw [←hx.2, compl_compl],
+    exact IH o o_lt_i x hx.1 },
+  rw ← hf,
+  exact generate_measurable.union (λn, (f n).val) typf
+end
+
+theorem generate_measurable_eq_gen_measurable :
+  {t | generate_measurable s t} = gen_measurable s :=
+begin
+  ext t, refine ⟨λ ht, _, λ ht, _⟩,
+  { have om1_nonz : ω₁ ≠ 0,
+    { unfold omega_1, exact ne_zero_of_out_nonempty _,},
+    induction ht with u hu u hu IH f hf IH,
+    exacts
+    [ self_subset_sigma0 s ω₁ om1_nonz hu,
+      empty_mem_sigma0 s ω₁ om1_nonz,
+      compl_mem_gen_measurable s u IH,
+      Union_mem_gen_measurable s IH ] },
+  { exact generate_measurable_of_mem_sigma0 s ω₁ t ht }
+end
+
+end gen_measurable
+
+section card_gen_measurable
```

Comment:
Since you haven't removed this yet from the PR, I assume it's not ready for review yet. But then you shouldn't have changed the tag back to "awaiting-review"...

Reply:
Oh, I'm sorry! I need to get a better acquaintance with how the communication works in this framework.
I thought that changing the tag was the way to signal that I had made changes. 

I am also not sure if my marking several comments as "resolved" is adequate.

Perhaps now it is the right time to change the label back to "awaiting-review", but this time I'll wait for your confirmation. Thanks for your patience :pray:.

Reply:
No problem, don't worry! Setting comments as resolved when you think you have dealt with them in an appropriate way is perfect: only keep open the ones for which it is not clear for you how to proceed. The tag change is indeed meant to say that you have dealt with all the comments, and that you think the PR is ready for another round of review. So please go ahead and change it back if you think everything is ready.

One warning: when the branch is merged into master, it will be erased. So make sure that you have a copy of the cardinality bounds in another branch, unless you want to prove it from scratch again!

Reply:
I think now it's ready for proper code review!

Reply:
I have some doubts that arouse while trying to apply this code to the case of topological space, and that's why I eliminated the `awaiting-review` tag.

I anticipate the following problem. When I instantiate the hierarchy on a second countable space, I think it would be better to have new definitions of `sigma0` and `pi0` (in another namespace) that include a countable basis as the `s` argument. Perhaps this requires defining things using typeclasses.

I'll try to make the query precise and post it at Zulip.